### PR TITLE
[OAP-CACHE][OAP-1748][POAE7-462] Enable externalDB to store CacheMetaInfo branch master

### PR DESCRIPTION
**Pull request by offthewall123**
## What changes were proposed in this pull request?

Enable externalDB to store CacheMetaInfo

## How was this patch tested?

E2E

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

File: oap-cache/oap/src/main/spark3.0.0/scala/org/apache/spark/sql/execution/datasources/FilePartition.scala [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    // Takes the first 3 hosts with the most data to be retrieved
+    val hdfsPreLocs = hostToNumBytes.toSeq.sortBy {
+      case (host, numBytes) => numBytes
+    }.reverse.take(3).map {
+      case (host, numBytes) => host
+    }.toArray
+
+    if (SparkEnv.get.conf.get(OapConf.OAP_EXTERNAL_CACHE_METADB_ENABLE) == true) {
```
Comment (offthewall123) [inline oap-cache/oap/src/main/spark3.0.0/scala/org/apache/spark/sql/execution/datasources/FilePartition.scala:50]: Here's the difference from origin FilePartition.scala

**Comment by offthewall123** [comment]
@jikunshang @winningsix @yma11 @yeyuqiang Please take a review thanks.

File: oap-cache/oap/src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/ParquetDataFiberReaderWriter.scala [hunk 2]
```diff
@@ -41,10 +42,15 @@ import org.apache.spark.unsafe.Platform
  */
 object ParquetDataFiberWriter extends Logging {
 
-  def dumpToCache(column: OnHeapColumnVector, total: Int): FiberCache = {
+  def dumpToCache(column: OnHeapColumnVector, total: Int, fiberId: FiberId = null,
+                  path: String = null, mc: ColumnChunkMetaData = null): FiberCache = {
     val header = ParquetDataFiberHeader(column, total)
     logDebug(s"will dump column to data fiber dataType = ${column.dataType()}, " +
-      s"total = $total, header is $header")
+      s"total = $total, header is $header")3
```
Comment (jikunshang) [inline oap-cache/oap/src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/ParquetDataFiberReaderWriter.scala:49]: typo?

**Review by jikunshang — CHANGES_REQUESTED** [review]
CI failed, please fix it.

File: oap-cache/oap/src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/ParquetDataFiberReaderWriter.scala (hunk 2 above)
Comment (offthewall123) [reply oap-cache/oap/src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/ParquetDataFiberReaderWriter.scala]: fixed

File: oap-cache/oap/src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/OapCache.scala [hunk 3]
```diff
@@ -1010,6 +1017,19 @@ class ExternalCache(fiberType: FiberType) extends OapCache with Logging {
     }
   }
 
+  def reportCacheMeta(fiberId: FiberId): Unit = {
+    fiberId match {
+      case binary: BinaryDataFiberId => binary
```
Comment (jikunshang) [inline oap-cache/oap/src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/OapCache.scala:1022]: can we merge case 1 and case 2 here?

**Review by jikunshang — APPROVED** [review]
LGTM  +1.

File: oap-cache/oap/src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/OapCache.scala (hunk 3 above)
Comment (offthewall123) [reply oap-cache/oap/src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/OapCache.scala:1022]: `BinaryDataFiberId` and `VectorDataFiberId` get report info by different ways, so i think can't merge it.

Comment (jikunshang) [reply oap-cache/oap/src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/OapCache.scala:1022]: I mean like this:
`case fiberId: Binary | Vector => fiberId.report()`
But seems scala doesn't support, see [link](https://stackoverflow.com/questions/1837754/match-multiple-cases-classes-in-scala).
Anyway, just a suggestion.

File: oap-cache/oap/src/main/scala/org/apache/spark/sql/internal/oap/OapConf.scala [hunk 4]
```diff
@@ -535,4 +535,28 @@ object OapConf {
       .doc("The commit algorithm version used by create index job")
       .intConf
       .createWithDefault(2)
+
+  val OAP_EXTERNAL_CACHE_METADB_ENABLE = {
+    SqlConfAdapter.buildConf("spark.sql.oap.external.cache.metaDB.enable")
```
Comment (HongW2019) [inline oap-cache/oap/src/main/scala/org/apache/spark/sql/internal/oap/OapConf.scala:540]: @offthewall123  To sync with Spark config style, we'd better modify .enable to .enabled,  PR #1802 and #1805 have made corresponding changes, pls pay attention when manually setting it in spark-defaults.conf.

Comment (offthewall123) [reply oap-cache/oap/src/main/scala/org/apache/spark/sql/internal/oap/OapConf.scala:540]: Got it.

Comment (zhixingheyi-tian) [reply oap-cache/oap/src/main/scala/org/apache/spark/sql/internal/oap/OapConf.scala:540]: @offthewall123 

Would you avoid overriding Spark Source code FilePartition.scala?
Now FilePartition.scala is the only one we have not decoupled from Spark Source.

Comment (offthewall123) [reply oap-cache/oap/src/main/scala/org/apache/spark/sql/internal/oap/OapConf.scala:540]: @zhixingheyi-tian I think can decouple it, will submit a refactor pr.

**Pull request merged**